ecopod/sendRawTransactionConditional.md

```diff
@@ -0,0 +1,53 @@
+
+**Note:** As a result of [EIP-3074](https://eips.ethereum.org/EIPS/eip-3074), the interop project is also exploring external validation as a mechanism to preventing DoS attack. A low-stakes scoped version of this for `eth_sendRawTransactionConditional` is a good opportunity for learning.
+
+
+# Problem Statement + Context
+
+Bundlers aggregate UserOps from an external mempool into a single transaction sent to the enshrined 4337 `Entrypoint` contract. It is the Bundler's responsibility to ensure every UserOp in the transaction will not revert when its validation step is executed. In between tx submission and inclusion, there is a chance network state changes causes the transaction to revert, even after local bundler validation.
+
+With private Bundler mempools, the likelihood of these reverts are small. However shared 4337 mempools are coming into [production](https://medium.com/etherspot/decentralized-future-erc-4337-shared-mempool-launches-on-ethereum-b6c860072f41), launched on Ethereum and some L2 testnets, decentralizing 4337 infrastructure. This is possible with no change on Ethereum through special block builders like Flashbots which support atomic transaction inclusion. An endpoint like `eth_sendRawTransactionConditional` is required to launch a shared mempools on L2 as there's only a single block builder, the sequencer, and the likelihood of reverts with a shared mempool would be too high of a cost for bundlers to operate.
+
+
+# Alternatives Considered
+
+## Continue with private 4337 mempools
+
+4337 account abstraction is live on optimism. Dapps likely utilize the bundler endpoints that are come with a vendor-specific SDK -- Alchemy, Pimilico, Thirdweb, etc. As 4337 infrastructure becomes more permissionless, we will likely later have to play catch-up to ensure the OP-Stack remains compatible while other L2 offerings have already moved towards supporting `eth_sendRawTransactionConditional`.
+
+## Verticalize the OP-Stack
+
+Rather than externalize the 4337 mempool, the op-stack could natively offer a UserOp mempool alongside the regular mempool, avoiding conflicts as the sequencer can ensure the bundled UserOps do not conflict with the latest chain state when sealing a new block. However, this adds additional complexity to the stack, where the proposer-builder seperation in 4337 nicely keeps account abstraction out of protocol. A chain operator can still choose to verticalize in proposed solution solution by allowlisting only its own bundler through `eth_sendRawTransactionConditional`, achieving the same outcome.
+
+
+# Proposed Solution
+
+1. Implement `eth_sendRawTransactionConditional` as described in the [spec](https://notes.ethereum.org/@yoav/SkaX2lS9j). A draft implementation for this [exists](https://github.com/ethereum/go-ethereum/compare/master...tynes:go-ethereum:eip4337)today. The conditional is applied on submission and tx inclusion.
```

Comment:
In a PGA, transactions are reprioritized and re-executed rapidly until the 2s sealing time is up.

Are conditionals based only on information from already-built blocks, or are they also able to access state that was modified by an earlier tx in the same block? I assume it's the former, otherwise there are some strange behaviors when filtering on re-execution. But wanted to make sure my understanding is correct.

Reply:
For the invariant to be held true, they must be evaluated against the pending state, ie the partially built block

Reply:
If invariants can come from partially built blocks, that creates some emergent behavior that doesn't sound great for UX:

Consider an invariant which is false on the first execution of an in-progress block. The transaction would get filtered out. But then if the invariant turns to *true* on recommit (because reprioritization changed the inclusion or outcome of the condition), the transaction has already been filtered out, and therefore still won't be included. Customers would see that the published block does not include their transaction, but *does* satisfy the invariant. In this way, invariants are less like simple `if` statements, and more like `never false` statements -- if the invariant filters the Tx even once, it wont' be included.

This is besides the extra compute pressure we'd be getting from checking these invariants on every recommit, which I want to be careful about.

Reply:
invariants are not applied on partially built blocks. Only the latest unsafe head. There is a chance that two conflicting bundler txs pass the mempool as they would be valid at the time of submission.

Based on the priority fee, one of them would get included and the other would then be dropped when the block is actually built.

+1 on the compute pressure, I added that as a risk. We could be smart in validation and reject the second competing bundler transactions that's detected to have a conflicting conditional, even when valid against the unsafe head but I think that's be overengineering to start and also introduces new behavior relative to the priority fee

Reply:
+1 on the extra compute pressure. There's an added section in the risks for this.

> The transaction would get filtered out. But then if the invariant turns to true on recommit (because reprioritization changed the inclusion or outcome of the condition)

This is a good point, the number of filtered out txs in the block building phase is something we need to keep metrics on. Would be interesting to see the metrics on arb which has this endpoint publicly available. With the usage being relatively low, my thinking is we can observe this behavior in testnet and be a bit reactionary with the safeguards present to rollback/shutoff the endpoint.